arch/common/shared_irq.c

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright 2023 NXP
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/sw_isr_table.h>
+#include <zephyr/arch/cpu.h>
+#include <zephyr/spinlock.h>
+
+struct shared_irq_client {
+	void (*routine)(const void *arg);
+	const void *arg;
+};
+
+struct shared_irq_data {
+	bool enabled;
+	struct shared_irq_client clients[CONFIG_SHARED_IRQ_MAX_NUM_CLIENTS];
+	uint32_t client_num;
+	struct k_spinlock lock;
+};
+
+static struct shared_irq_data shared_irq_table[CONFIG_NUM_IRQS];
```

Comment:
This is an awfully big table!  8 bytes per record, 16 (default) records per IRQ, and most platforms have a few dozen IRQ slots.  At the very least I'd say 16 is way too generous for a MAX_NUM_CLIENTS default.  The much more typical number of drivers forced to share an interrupt is "two". :)

Reply:
Actually, you'd pay for it with a little code size, but one clever way to do this would be to put a dlist node into the structs and keep them in lists, "allocating" them pool-style (i.e. walk the array linearly to start, keep deleted records in a free list, and allocate from there when the original N have been exhausted).  Then you'd only need one record (albeit one of 16 bytes instead of 8) for each shared handler of any IRQ in the system.